File: implementations/node/lib/pipeline.js
/** Pipeline - orchestrates the processing chain */

import { Cleaner } from './cleaner.js';
import { Chunker } from './chunker.js';
import { EntityEngine } from './entities.js';
import { RelationEngine } from './relations.js';
import { GraphEngine } from './graph.js';
import { InsightsEngine } from './insights.js';

export class Pipeline {
  constructor(config = {}) {
    this.cleaner = new Cleaner(config);
    this.chunker = new Chunker(config);
    this.entityEngine = new EntityEngine(config);
    this.relationEngine = new RelationEngine(config);
    this.graphEngine = new GraphEngine(config);
    this.insightsEngine = new InsightsEngine(config);
  }

  extractFromText(text) {
    const cleanedText = this.cleaner.clean(text);
    const chunks = this.chunker.chunk(cleanedText);
    const entities = this.entityEngine.extract(cleanedText);
    const relations = this.relationEngine.extract(entities, chunks);
    const graph = this.graphEngine.build(entities);
    const insights = this.insightsEngine.compute(entities, chunks, cleanedText);

    return this._buildResult("", "", cleanedText, chunks, entities, relations, graph, insights);
  }

  _buildResult(url, title, text, chunks, entities, relations, graph, insights) {
    const sortedEntities = [...entities].sort((a, b) => {
      if (a.type !== b.type) return a.type.localeCompare(b.type);
      return a.value.localeCompare(b.value);
    });

    const sortedChunks = [...chunks].sort((a, b) => a.index - b.index);

    const sortedRelations = [...relations].sort((a, b) => {
      if (a.source !== b.source) return a.source.localeCompare(b.source);
      return a.target.localeCompare(b.target);
    });

    return {
      meta: { title, url },
      content: { text },
      chunks: sortedChunks,
      entities: sortedEntities,
      relations: sortedRelations,
      graph: {
        nodes: [...graph.nodes].sort((a, b) => a.id.localeCompare(b.id)),
        edges: [...graph.edges].sort((a, b) => {
          if (a.source !== b.source) return a.source.localeCompare(b.source);
          return a.target.localeCompare(b.target);
        })
      },
      insights
    };
  }
}

File: implementations/dart/lib/pipeline.dart
import 'dart:convert';
import 'dart:collection';
import 'cleaner.dart';
import 'chunker.dart';
import 'entities.dart';
import 'relations.dart';
import 'graph.dart';
import 'insights.dart';

class Pipeline {
  final Cleaner _cleaner = Cleaner();
  final Chunker _chunker = Chunker();
  final Entities _entities = Entities();
  final Relations _relations = Relations();
  final Graph _graph = Graph();
  final Insights _insights = Insights();

  Map<String, dynamic> extractFromText(String text) {
    final cleanedText = _cleaner.clean(text);
    final chunkList = _chunker.chunk(cleanedText);
    final entityList = _entities.extract(cleanedText);
    final relationList = _relations.extract(entityList);
    final graphData = _graph.build(entityList);
    final insightsData = _insights.compute(entityList, chunkList, cleanedText);

    return buildResult(
      cleanedText,
      chunkList,
      entityList,
      relationList,
      graphData,
      insightsData,
    );
  }

  Map<String, dynamic> buildResult(
    String text,
    List<Chunk> chunks,
    List<Entity> entities,
    List<Relation> relations,
    GraphData graphData,
    Map<String, dynamic> insightsData,
  ) {
    final sortedEntities = List<Entity>.from(entities)
      ..sort((a, b) {
        final typeCmp = a.type.compareTo(b.type);
        if (typeCmp != 0) return typeCmp;
        return a.value.compareTo(b.value);
      });

    final chunksList = chunks.map((c) => c.toMap()).toList();
    final entitiesList = sortedEntities.map((e) => e.toMap()).toList();
    final relationsList = relations.map((r) => r.toMap()).toList();

    return LinkedHashMap<String, dynamic>.from({
      'meta': LinkedHashMap<String, dynamic>.from({'title': '', 'url': ''}),
      'content': LinkedHashMap<String, dynamic>.from({'text': text}),
      'chunks': chunksList,
      'entities': entitiesList,
      'relations': relationsList,
      'graph': graphData.toMap(),
      'insights': insightsData,
    });
  }

  String toJson(Map<String, dynamic> map) {
    return const JsonEncoder.withIndent('  ').convert(map);
  }
}
